system/jlib/jptree.cpp

```diff
@@ -3176,6 +2813,261 @@ bool PTree::checkPattern(const char *&xxpath) const
     return ret;
 }
 
+AttrValue *PTree::findAttribute(const char *key) const
+{
+    AttrValue *a = attrs+numAttrs;
+    if (isnocase())
+    {
+        while (a-- != attrs)
+        {
+            if (strieq(a->key->get(), key))
+                return a;
+        }
+    }
+    else
+    {
+        while (a-- != attrs)
+        {
+            if (streq(a->key->get(), key))
+                return a;
+        }
+    }
+    return nullptr;
+}
+
+const char *PTree::getAttributeValue(const char *key) const
+{
+    AttrValue *e = findAttribute(key);
+    if (e)
+        return e->value->get();
+    return nullptr;
+}
+
+unsigned PTree::getAttributeCount() const
+{
+    return numAttrs;
+}
+
+AttrValue *PTree::getNextAttribute(AttrValue *cur) const
+{
+    if (0 == numAttrs)
+        return nullptr;
+    else if (nullptr == cur)
+        return attrs;
+    else
+    {
+        if (cur == (attrs+(numAttrs-1)))
+            return nullptr;
+        return ++cur;
+    }
+}
+
+
+//////////////////////
+
+// LocalPTree
+
+LocalPTree::LocalPTree(const char *_name, byte _flags, IPTArrayValue *_value, ChildMap *_children) : PTree(_flags, _value, _children)
+{
+    if (_name)
+        setName(_name);
+}
+
+LocalPTree::~LocalPTree()
+{
+    if (name)
+        free(name);
+    if (!attrs)
+        return;
+    AttrValue *a = attrs+numAttrs;
+    while (a--!=attrs)
+    {
+        free(a->key);
+        free(a->value);
+    }
+    free(attrs);
+}
+
+void LocalPTree::setName(const char *_name)
+{
+    HashKeyElement *oname = name;
+    if (!_name)
+        name = nullptr;
+    else
+        name = AtomRefTable::createKeyElement(_name, isnocase());
+    if (oname)
+        free(oname);
+}
+
+bool LocalPTree::removeAttribute(const char *key)
+{
+    AttrValue *del = findAttribute(key);
+    if (!del)
+        return false;
+    numAttrs--;
+    unsigned pos = del-attrs;
+    memmove(attrs+pos, attrs+pos+1, (numAttrs-pos)*sizeof(AttrValue));
+    return true;
+}
+
+void LocalPTree::setAttribute(const char *key, const char *val)
+{
+    if (!key)
+        return;
+    if (!validateXMLTag(key+1))
+        throw MakeIPTException(-1, "Invalid xml attribute: %s", key);
+    if (!val)
+        val = "";  // cannot have NULL value
+    AttrValue *v = findAttribute(key);
+    if (v)
+    {
+        if (streq(v->value->get(), val))
+            return;
+        AttrStrC::destroy((AttrStrC *)v->value);
+        v->value = AttrStrC::create(val);
+    }
+    else
+    {
+        attrs = (AttrValue *)realloc(attrs, (numAttrs+1)*sizeof(AttrValue));
+        if (isnocase())
```

Comment:
future: I suspect there is some scope for moving functions from AttrStrNC/C into AttrStr - e.g., create(key, isnocase()) and hash/destroy.